lib/devise-security/validators/password_complexity_validator.rb

```diff
@@ -19,17 +19,29 @@ class DeviseSecurity::PasswordComplexityValidator < ActiveModel::EachValidator
   }.freeze
 
   def validate_each(record, attribute, value)
-    active_pattern_keys.each do |key|
-      minimum = [0, options[key].to_i].max
+    active_pattern_keys(record).each do |key|
+      minimum = [0, constraints(record).fetch(key).to_i].max
       pattern = Regexp.new PATTERNS[key]
 
-      unless (value || '').scan(pattern).size >= minimum
+      unless value.to_s.scan(pattern).size >= minimum
         record.errors.add attribute, :"password_complexity.#{key}", count: minimum
       end
     end
   end
 
-  def active_pattern_keys
-    options.keys & PATTERNS.keys
+  def constraints(record)
+    # TODO - there are examples where there is and isn't a `constraints`. Is one format deprecated?
+    constraints = options[:constraints] || options
```

Comment:
Based on the other test changes, it looks like maybe you were moving towards requiring the `constraints` key. Would this be a breaking change?